# Student Performance Indicator

### Life cycle of Machine learning Project

- Understanding the Problem Statement
- Data Collection
- Data Checks to perform
- Exploratory data analysis
- Data Pre-Processing
- Model Training
- Choose best model

## 1) Problem statement

- This project understands how the student's performance (test score) is affected by other variables such as Gender, Ethnicity, Parental level of education, Lunch and Test preparation course.

## 2) Data Collection

- Data Source - https://www.kaggle.com/datasets/spscientist/students-performance-in-exams?select=StudentsPerformance.csv
- The data consists of 8 column and 1000 rows.

## 2.1 Import Data and Required Packages

Importing Pandas, Numpy, Matplotlib, Seaborn and Warnings Library.

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

## Import the CSV Data as Pandas Dataframe

In [2]:
df = pd.read_csv('data/stud.csv')

### Show top 5 rows

In [3]:
df.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

### Shape of the data

In [4]:
df.shape

(1000, 8)

## 2.2 Data Information

- gender: sex of students -> (Male/Female)
- race/ethnicity:ethnicity of students -> (Group A, B, C, D, E)
- Parental level of education: parents' final education -> (bachelor's degree, some college, master's degree, associate's degree, high school
- lunch: having lunch before test (standard or free/reduced)
- test preparation cource: complete or not complete before test
- math score
- reading score
- writing score

## 3. Data Checks to perform

- Check Missing values
- Check Duplicates
- Check Data type
- Check number of unique value of each column
- Check statistics of data set
- Check various categories present in the different categorical column

## 3.1 Check Missing Values

In [5]:
df.isnull().sum()

gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

#### There are no missing values in the data set

## 3.2 Check Duplicates

In [6]:
df.duplicated().sum()

0

#### There are no duplicated values in the data set

## 3.3 Check Data types

In [7]:
# Check null and Data types
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


## 3.4 Checking the number of unique values of each column

In [9]:

df.nunique()

gender                          2
race/ethnicity                  5
parental level of education     6
lunch                           2
test preparation course         2
math score                     81
reading score                  72
writing score                  77
dtype: int64

## 3.5 Checking statistics of data set

In [10]:
df.describe()

math score  reading score  writing score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

### Insight

- From above description of numerical data, all means are very close to each other- between 66 and 68.05;
- All standard deviations are also close - between 14.6 and 15.19
- While there is a minimum score 0 for math, for writingminimum much higher = 10 and for reading much higher = 17

## 3.7 Exploring Data

In [11]:
print("Categories in 'gender' variable: ",end=' ')
print(df['gender'].unique())

print("Categories in 'race/ethnicity' variable: ",end=' ')
print(df['race/ethnicity'].unique())

print("Categories in 'parental level of eduction' variable: ", end=" ")
print(df['parental level of education'].unique())

print("Categories in 'lunch' variable: ", end=" ")
print(df['lunch'].unique())

print("Categories in 'test preparation course' variable: ")
print(df['test preparation course'].unique())

Categories in 'gender' variable:  ['female' 'male']
Categories in 'race/ethnicity' variable:  ['group B' 'group C' 'group A' 'group D' 'group E']
Categories in 'parental level of eduction' variable:  ["bachelor's degree" 'some college' "master's degree" "associate's degree"
 'high school' 'some high school']
Categories in 'lunch' variable:  ['standard' 'free/reduced']
Categories in 'test preparation course' variable: 
['none' 'completed']


In [17]:
# define numerical and categorical columns

numeric_features = [feature for feature in df.columns if df[feature].dtype != 'object']
categorical_feature = [feature for feature in df.columns if df[feature].dtype == 'object']

# print column
print('we have {} numerical feature: {}'.format(len(numeric_features),numeric_features))
print('we have {} categorical feature: {}'.format(len(categorical_feature),categorical_feature))

we have 3 numerical feature: ['math score', 'reading score', 'writing score']
we have 5 categorical feature: ['gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course']


## 3.8 Adding columns for "Total Score" and "Average"

In [18]:
df['total_score'] = df['math score'] + df['reading score'] + df['writing score']
df['average'] = df['total_score']/3
df.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  \
0                    none          72             72             74   
1               completed          69             90             88   
2                    none          90             95             93   
3                    none          47             57             44   
4                    none          76             78             75   

   total_score    average  
0          218  72.666667  
1          247  82.333333  
2          278  92.666667  
3          148  49.333333  
4          229  76.333333

In [20]:
reading_full = df[df['reading score'] ==100]['average'].count()
writing_full = df[df['writing score'] == 100]['average'].count()
math_full = df[df['math score'] == 100]['average'].count()

print(f'Number of student obtain full mark in maths: {math_full}')
print(f'Number of student obtain full marks in writing: {writing_full}')
print(f'Number of student obtain full marks in reading: {reading_full}')

Number of student obtain full mark in maths: 7
Number of student obtain full marks in writing: 14
Number of student obtain full marks in reading: 17
